Add worklist subscription and filters

paulfalgout opened the pull request:
Shortcut Story ID: [sc-59124]

So to implement the filters messages we have to grab a model from the store.. rather than copy this pattern to a bunch of places I moved it to the entities channel.

Then I looked at other places where this pattern existed and changed it to `Model.getRelationship` which will coincide eventually with simplify the various relationships once https://github.com/RoundingWell/care-ops-frontend/pull/1320 is complete.

For not the filters implementation on the ws service is overly simplistic, but I don't know what kind of filter maintenance we're going to run into yet or if it's always a simple replace.

But I'm opening this as a draft as I think I need to real-world test the socket filters before adding tests in cypress etc.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Enhanced the patient worklist with real-time notifications that immediately reflect updates for flows and actions.
  - Introduced improved filtering for a more accurate and responsive display of data.
  - Streamlined subscription mechanism to accommodate filters during WebSocket communications.

- **Bug Fixes**
  - Strengthened error handling during live data communications, reducing the risk of processing incomplete updates.

- **Tests**
  - Added automated tests to ensure that real-time notifications trigger the correct UI updates for patient worklist changes.
  - Implemented specifications for socket notifications related to flows and actions to verify UI updates.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

paulfalgout committed:
Add simplistic filters to ws service

There’s probably a better way to do this, but might need more use cases first

paulfalgout committed:
Implement worklist socket subscription

paulfalgout committed:
Remove notifications that have no event data

Apparently Bref requires something be sent at point when there’s nothing to send.. so we’ll just exit out here.

But I can’t test it because it doesn’t seem possible to send nothing in the node websocket setup.

paulfalgout committed:
Filter out action flow_states from notifications

flow_states aren’t able to be filtered on the BE so we’re doing an extra filter here

nmajor25 committed:
Add `this._isFetching` boolean value to base model

Set to `true` when model is being fetched

In worklist app, we can use that to not do duplicate `model.fetch()` requests after ws notifications are received

nmajor25 committed:
Use `data.payload.resource` for flow `ResourceCreated` ws notifications on worklists

Instead of `data.resource`, which contains the patient data

nmajor25 commented on src/js/apps/patients/worklist/worklist_app.js:
```diff
@@ -161,14 +161,20 @@ export default App.extend({
+    this.listenTo(channel, 'message', ({ category, payload }) => {
+      if (category !== 'ResourceCreated') return;
+
+      const model = Radio.request('entities', 'get:store', payload.resource);
```
I wanted to do this ☝️ in the `onMessage()` of the `/js/services/ws.js` [file](https://github.com/RoundingWell/care-ops-frontend/blob/develop/src/js/services/ws.js#L101-L116).

Specifically I wanted to do something like:

```js
if (data.category === 'ResourceCreated') {
  model = Radio.request('entities', 'get:store', data.payload.resource);
  channel.trigger('message', data, model);
  return;
}
```

But that would break our handling of the `ResourceCreated` ws notification for actions on the patient flow page.

Both of these `ResourceCreated`'s have both a `data.resource` and `data.payload.resource`. So I'm not sure how to differentiate between the two in that `onMessage()` function.

nmajor25 commented on src/js/apps/patients/worklist/worklist_app.js:
```diff
@@ -161,14 +161,20 @@ export default App.extend({
 
         this.collection.add(model);
+
+        channel.request('add', model);
```
Tested this locally and it worked well 👍

nmajor25 commented on src/js/apps/patients/worklist/worklist_app.js:
```diff
@@ -161,14 +161,20 @@ export default App.extend({
         // notifications do not fully filter our action flow_states
         if (filterKey === 'actions' && !flowStates.includes(model.getFlow().getState().id)) return;
 
         this.collection.add(model);
```
Do we need to do anything with the `this.filteredCollection` and `this.editableCollection` here?

I haven't looked into the details yet to answer that question.

nmajor25 commented on src/js/base/model.js:
```diff
@@ -105,7 +105,22 @@ export default Backbone.Model.extend(extend({
+    if (!this.notifications) {
+      this.notifications = [data];
+      return;
+    };
```
Is there a way to set this automatically set `model.notifications: []` initially for each model? Do we want to do that?

If we did that we wouldn't have to check if `this.notifications` exists here. And just do the `push()`.

nmajor25 replied:
And wasn't sure what to use for naming. But went with `model.notifications` for now.

nmajor25 commented on src/js/apps/patients/worklist/worklist_app.js:
```diff
@@ -161,14 +161,21 @@ export default App.extend({
 
-    this.listenTo(channel, 'message', (message, model) => {
+    this.listenTo(channel, 'message', ({ category, resource, payload }) => {
+      if (!category) return;
```
In my local dev environment I'm getting a ws notification that contains a lambda error. When that happens this `listenTo` is called with `category`, `resource`, and `payload` as `undefined`.

I don't know if this is something that could happen in production or not.

<img width="833" alt="Screenshot 2025-02-24 at 4 50 48 PM" src="https://github.com/user-attachments/assets/5ab13ce5-f8ba-42e2-86b7-2d98b5c96436" />

paulfalgout replied:
we shouldn't need the `if` here.  should be able to handle those errors prior to triggering the message

nmajor25 replied:
I'm looking at that currently. Would be in the next commit. I have a `event.data.category` check added to the `onMessage()` [here](https://github.com/RoundingWell/care-ops-frontend/blob/develop/src/js/services/ws.js#L101C3-L101C12). 

So at the top of that function it would be:

```js
if (!event.data || !event.data.category) return;
```

But still need to test it locally and think about it a little more.

nmajor25 committed:
If model is currently being fetched, queue any incoming ws notification for that model. And then handle those ws notifications once the fetch finishes

nmajor25 commented on src/js/base/model.js:
```diff
@@ -105,7 +105,22 @@ export default Backbone.Model.extend(extend({
+    if (!this.notifications) {
+      this.notifications = [data];
+      return;
+    }
```
Is there a way to set `model.notifications: []` initially for each model? Do we want to do that?

If we did that we wouldn't have to check if `this.notifications` exists here. And just do the `push()`.

nmajor25 replied:
And wasn't sure what to use for naming. But went with `model.notifications` for now.

paulfalgout replied:
there is not a good place to instantiate this, but it should probably be `_notifications` and it should be `{}`. where `this.notifications['StateChanged']` would always be the latest `StateChanged`

nmajor25 replied:
`_notifications` naming makes sense.

But why an object `{}` instead of an array `[{}]`?

paulfalgout replied:
because we only need to run the most recent of a particular `category`

nmajor25 replied:
Ah okay, right. Using an object we can just replace the previous notification in the queue if one already exists with the same notification category (`StateChanged`, `OwnerChanged`, etc.).

paulfalgout commented on src/js/services/ws.js:
```diff
@@ -99,6 +103,9 @@ export default App.extend({
   onMessage(event) {
+    /* istanbul ignore next: Can't test this bref functionality in node websockets */
+    if (!event.data) return;
+
```
I think we should handle `event.data.error` near the pong return.. such that we trigger `error` instead of `message` and not have a `if (!category)` check in the message handler.

paulfalgout commented on src/js/base/model.js:
```diff
@@ -98,7 +105,22 @@ export default Backbone.Model.extend(extend({
+    this.notifications.push(data);
+  },
+  handleMessage(data) {
+    if (this.isFetching()) {
```
I think extending the attributes with the updated_at should happen here so that it's stored with the queue.. that timestamp is when the action occurred, not when it was processed.

paulfalgout commented on src/js/apps/patients/worklist/worklist_app.js:
```diff
@@ -151,6 +155,36 @@ export default App.extend({
+        this.collection.add(model);
+        channel.request('add', model);
+
+        if (model.notifications) {
```
we should either entirely handle this inside the base model, or entirely handle it here..  perhaps the base model is better..  I can't think of why we wouldn't always want this to happen

nmajor25 replied:
I think it would work to do this in the base model.

But does `this.collection.add(model)` need to happen before we handle the queued notifications? Or does that order not matter?

For example if the order of things happened like this, would the correct flow.state show in the UI?

1. A `StateChanged` notification is queued while a new flow is fetched.
2. Flow fetch finishes.
3. Queued `StateChanged` ws notification is handled and state is updated on the flow model.
4. Flow model is added to the worklist collection via `this.collection.add(model)`, which makes it show in the UI.

Does the order of 3 & 4 matter? My assumptions is that it doesn't matter. Either way the flow's state should show correctly in the UI.

paulfalgout replied:
better to do data before render

nmajor25 replied:
Yep 👍

paulfalgout commented on src/js/base/model.js:
```diff
@@ -36,13 +36,20 @@ export default Backbone.Model.extend(extend({
     // Resolve with entity if successful
     return fetcher.then(response => {
+      this._isFetching = false;
+
```
we should likely process the queue here.  and it's likely `_.each(this._notifications, this._handleMessage);` where `_handleMessage` is after the isFetching return

paulfalgout replied:
why `, data => { this._handleMessage(data) });` over `, this._handleMessage);`?

nmajor25 replied:
I was getting an error where `this` is `undefined`, is why I switched to that.

`each(this._notifications, this._handleMessage);` results in that error. But I just tested this and it works:       `each(this._notifications, () => { this._handleMessage });`.

paulfalgout replied:
ah yep, set the context `each(this._notifications, this._handleMessage, this);`

nmajor25 replied:
Nice that works, updated to that.

paulfalgout commented on src/js/base/model.js:
```diff
@@ -36,13 +36,26 @@ export default Backbone.Model.extend(extend({
     return fetcher.then(response => {
+      this._isFetching = false;
+
+      if (this._notifications) {
```
no need for the if..  I think `each(this._notifications, this._handleMessage);` is fine here.

nmajor25 replied:
Yep, not needed 👍

paulfalgout commented on src/js/base/model.js:
```diff
@@ -98,12 +109,30 @@ export default Backbone.Model.extend(extend({
   },
-  handleMessage({ category, resource, author, payload }) {
-    payload.attributes = extend({}, payload.attributes, { updated_at: dayjs.utc().format() });
+  queueMessage(data) {
```
So this function looks complex to me, but I get why.. but we could start using some of the newer operators to make this a bit simpler:

```js
queueMessage(data) {
 return (this._notifications ||= {})[data.category] = data;
}
```

nmajor25 replied:
That's pretty cool, didn't know `||=` existed until now 👍 

But `vite-plugin-eslint` doesn't support it. It's using eslint `v8.21.0` ([link](https://github.com/gxmari007/vite-plugin-eslint/blob/main/package.json#L50)). Eslint added support logical assignment operators in `v8.24.0` ([source](https://eslint.org/docs/latest/rules/logical-assignment-operators#version)).

That package's repo seems to be abandoned, so we probably want to switch to something different.

paulfalgout replied:
That's just the dev dependency.  We're using 9 and it's a peer dependency https://github.com/gxmari007/vite-plugin-eslint/blob/main/package.json#L60

paulfalgout replied:
FWIW it looks like the vite community doesn't run eslint as part of the build process.  They're using it in pre-commit pre-build hooks..  I find that mildly annoying

shadowhand replied:
FWIW backend runs static analysis (including lint) on commit and on test (CI).

paulfalgout commented on src/js/apps/patients/worklist/worklist_app.js:
```diff
@@ -151,6 +154,29 @@ export default App.extend({
+
+      if (this.collection.get(model) || model.isFetching()) return;
+
+      model.fetch().then(() => {
```
I think this'll be:

```
const include = [
      'program-action.program',
      'flow.program-flow.program',
    ].join()

const data = {
  fields: { flows: ['name', 'state'] },
  include,
};

model.fetch({ data }).then(...
```

Honestly that's probably enough logic to move it to a `Radio.request('entities', 'fetch:action..`

Or maybe just add the fields to `'fetch:actions:model'` and use that over model.fetch here.

nmajor25 replied:
Great thanks, I just got this working in my local dev environment. 

For now I'm just doing the following:

```js
  // js/apps/patients/worklist/worklist_app.js

  _getWsFetchOptions(filterKey) {
    if (filterKey === 'flows') return {};

    return {
      data: {
        fields: { flows: ['name', 'state'] },
        include: ['program-action.program', 'flow.program-flow.program'].join(),
      },
    };
  },
  subscribe() {

    . . .

    const options = this._getWsFetchOptions(filterKey);

    model.fetch(options).then(() => {

       . . .

    })
  },
```

nmajor25 replied:
Having this working now allows me to finish creating the initial tests for the action version of the worklist. After that I'll come back and do some refactoring on this and other parts too.

paulfalgout commented on src/js/apps/patients/worklist/worklist_app.js:
```diff
@@ -151,6 +154,29 @@ export default App.extend({
 
     this.showList();
   },
+  subscribe() {
```
Alright I think this whole function needs refactoring..

we should probably just split based on `filterKey` and hardcode types for this one..

also somewhere we can call break this up into another function
```
subscribe() {
  this.listenTo(... () => {
    fetch().then(() => {. // <- one or both of these could be a separate handler..
```

But I think we still want to check the store for the resource.. but if we need to fetch it, as mentioned below.. switch to either the action or flow entity request rather than adding the fetch specifications here for include/fields.

nmajor25 replied:
I took a crack at refactoring this in https://github.com/RoundingWell/care-ops-frontend/pull/1380/commits/836c7c01aa9d8cd6d86bfbacb59c0977853f5302, let me know what you think.

nmajor25 committed:
If notification doesnt have a category, dont trigger message

So the app doesnt handle ws notifications that are errors, only actual entity events

nmajor25 committed:
Include flow data in api response when fetching new ws action model

nmajor25 committed:
Add tests for web sockets on flow & action versions of worklist

nmajor25 commented on test/integration/patients/worklist/worklist.js:
Summary of what I'm testing here.

**Flow list notifications**:

* Load owned-by worklist with one flow
    * Test each ws notification category that applies on preloaded flow
        * NameChanged
            * Verify updated_at date
        * OwnerChanged
        * StateChanged
    * Add flow to list via `ResourceCreated` ws notification (w/ trailing `StateChanged` msg)
        * Verify included is not in request URL string
        * Verify flow count
        * Verify flow was added to list
        * Verify trailing `StateChanged` was queued and processed after model fetch finishes
        * Verify new flow is subscribed to future ws notifications
    * Add new flow to list via `StateChanged` ws notification
    * Add new flow to list via `OwnerChanged` ws notification

**Action list notifications**:
* Load owned-by worklist with one action
    * Test each ws notification category that applies on preloaded action
        * NameChanged
            * Verify updated_at date
        * OwnerChanged
        * ActionDueChanged
        * DetailsChanged
        * StateChanged
        * CommentAdded
        * AttachmentAdded
        * SharingUpdated
    * Add action to list via `ResourceCreated` ws notification (w/ trailing `StateChanged` msg)
        * Verify included  data is in request URL string
        * Verify action count
        * Verify action was added to list
        * Verify trailing `StateChanged` was queued and processed after model fetch finishes
        * Verify new action is subscribed to future ws notifications
    * Add new action to list via `StateChanged` ws notification
    * Add new action to list via `OwnerChanged` ws notification
    * Handle ws notification for new action with a flow state not in the current filters
      * In this scenario, the action should not be added to list

paulfalgout commented on test/integration/patients/worklist/worklist.js:
```diff
@@ -303,6 +303,287 @@ context('worklist page', function() {
+      id: 'owned-by',
+      listType: 'flows',
+      flowsSortId: 'sortCreatedDesc',
+      clinicianId: '11111',
```
Don't love using new string ids.. is this `getCurrentClinician().id`?

nmajor25 replied:
Yep, that's the current clinician (i.e. `getCurrentClinician()`). I updated the tests to reference that directly instead of using `11111`.

paulfalgout commented on test/integration/patients/worklist/worklist.js:
```diff
@@ -303,6 +303,287 @@ context('worklist page', function() {
+        id: testPatient1.id,
+      },
+      payload: {
+        resource: {
```
I do badly want destructing within an object like
```s
resource: { id, type } = testPatient1, 
```
but alas none of these types of proposals go anywhere right now in TC39 😿 

we could make some sort of generic helper. `function foo({ id, type ]) => { id, type };`
`resource: foo(testNewSocketFlow)`

thought for another day

nmajor25 replied:
Ya that would be pretty nice 👍

paulfalgout commented on test/integration/patients/worklist/worklist.js:
```diff
@@ -848,6 +1129,445 @@ context('worklist page', function() {
+      relationships: {
+        state: getRelationship(stateTodo),
+        flow: getRelationship(testFlow),
+        files: getRelationship([]),
```
it's ok to do so, but I suspect we don't need to set a lot of this for the tests?  details?  due_time? duration. files relationship?  If they're just random values from `getAction` does that potentially cause flake?  Ideally were setting whatever minimum test data we need and letting `getAction` do the rest

nmajor25 replied:
Agreed, I cleaned up a number of these: https://github.com/RoundingWell/care-ops-frontend/compare/003a41f45cc3b2abdda8b3aa33189122edb965a0..54708cca95f3c34880bfa2975ca936e7e887260c

paulfalgout commented on test/integration/patients/worklist/worklist.js:
```diff
@@ -848,6 +1129,445 @@ context('worklist page', function() {
+      .routeFormByAction()
+      .routeFormDefinition()
+      .routeLatestFormResponse()
+      .visitOnClock('/worklist/owned-by', { now: testTs() })
```
the visitOnClock is to set the time to now or why is this needed?

nmajor25 replied:
It's to avoid flakiness when checking that the `updated_at` timestamp is updated correctly in the UI.

First introduced in this commit: https://github.com/RoundingWell/care-ops-frontend/commit/8eb7d817796c7852cc0ce7842fc676c38ae23017.

paulfalgout reviewed (commented):
looking pretty good.

nmajor25 committed:
Refactor `subscribe()` function in worklist app

paulfalgout commented on src/js/apps/patients/worklist/worklist_app.js:
```diff
@@ -151,6 +154,51 @@ export default App.extend({
+
+      this.collection.add(model);
+
+      Radio.channel('ws').request('add', model);
```
`Radio.request('ws', 'add',...` here and above

nmajor25 replied:
Updated 👍

paulfalgout commented on src/js/apps/patients/worklist/worklist_app.js:
```diff
@@ -151,6 +154,51 @@ export default App.extend({
+        return;
+      }
+
+      if (filterKey === 'actions') {
```
I think you'll need to remove the `if` here and have a default, otherwise we'll have an uncovered `else` branch

nmajor25 replied:
Updated 👍

paulfalgout reviewed (commented):
couple of small things

nmajor25 commented:
@paulfalgout coveralls coverage looks good 👍

nmajor25 commented on src/js/base/model.js:
```diff
@@ -36,13 +36,22 @@ export default Backbone.Model.extend(extend({
     // Model fetches default to aborting.
     const fetcher = Backbone.Model.prototype.fetch.call(this, extend({ abort: true }, options));
 
+    this._isFetching = true;
+
     // Resolve with entity if successful
     return fetcher.then(response => {
+      this._isFetching = false;
+
```
Correct me if I'm wrong here @paulfalgout, but I think we'd want `this._isFetching` to remain `true` in the scenario of a fetch error. Since we wouldn't want the model's queued ws notifications to be processed. So I think the code we current have is good.

paulfalgout replied:
hmm no I think we should add this catch with the rethrow of the error.  It is more correct for the state of the base model, regardless of how it affects us downstream.  if a fetch fails though, we're not necessarily doing anyhing in order to process the queue correctly anyhow.

nmajor25 replied:
That's been added in commit 0222306fd0fa389b896351ffa29ceb1c20e35305.

Nice catch @coderabbitai 👍

nmajor25 commented on src/js/apps/patients/worklist/worklist_app.js:
```diff
@@ -151,6 +154,51 @@ export default App.extend({
 
     this.showList();
   },
+  subscribe() {
+    const channel = Radio.channel('ws');
+    const filterKey = this.getState().getType();
+
+    channel.request('subscribe', this.collection.models, { filters: { [filterKey]: this.filters } });
+
+    this.listenTo(channel, 'message', ({ category, resource, payload }) => {
+      const modelResource = category === 'ResourceCreated' ? payload.resource : resource;
+      const model = Radio.request('entities', 'get:store', modelResource);
+
+      if (this.collection.get(model) || model.isFetching()) return;
+
+      if (filterKey === 'flows') {
+        this._fetchFlow(model);
+
+        return;
+      }
+
+      this._fetchAction(model);
+
+      return;
+    });
+  },
```
In `src/js/apps/patients/worklist/worklist_app.js` we now check that the data contains a `category`. And we don't handle that web socket notification if it doesn't have that.

If there's no category, that's a good indication that the data isn't going to be a valid web socket notification.

nmajor25 committed:
Add `.catch()` to base model fetch

When that happens, set `_isFetching = false` and throw the error

paulfalgout reviewed (commented):
approved on my end

paulfalgout commented:
The error is coveralls 504ing.
